src/test/cpp/CMakeLists.txt

```diff
@@ -1,18 +1,59 @@
 cmake_minimum_required(VERSION 3.9)
 
-project(test_cpp)
-
-add_executable(test_cpp "")
-target_sources(test_cpp PUBLIC pch.cpp meta_reader_filter.cpp meta_reader_cache.cpp cmd_reader.cpp simple.cpp)
-target_include_directories(test_cpp PUBLIC ${XLANG_LIBRARY_PATH} ..\\output\\foundation ..\\output\\simple)
+add_executable(cpp_test)
+target_include_directories(cpp_test BEFORE PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/implementation "${CMAKE_CURRENT_SOURCE_DIR}/generated")
+target_sources(cpp_test PUBLIC pch.cpp activation.cpp test/Async.cpp test/Collections.cpp test/Edge.cpp test/Fast.cpp "${CMAKE_CURRENT_SOURCE_DIR}/generated/module.g.cpp")
 
 if (MSVC)
-    TARGET_CONFIG_MSVC_PCH(test_cpp pch.cpp pch.h)
-    target_link_libraries(test_cpp windowsapp ole32)
+    TARGET_CONFIG_MSVC_PCH(cpp_test pch.cpp pch.h)
+    target_link_libraries(cpp_test windowsapp)
 else()
-    target_link_libraries(test_cpp c++ c++abi c++experimental)
-    target_link_libraries(test_cpp -lpthread)
+    target_link_libraries(cpp_test c++ c++abi c++experimental)
+    target_link_libraries(cpp_test -lpthread)
 endif()
 
-target_sources(test_cpp PUBLIC main.cpp)
-add_dependencies(test_cpp test_il)
+target_sources(cpp_test PUBLIC main.cpp)
+
+file(TO_NATIVE_PATH ${CMAKE_CURRENT_BINARY_DIR}/../../tool/cpp/cpp.exe cpp_exe)
+file(TO_NATIVE_PATH ${CMAKE_CURRENT_SOURCE_DIR} project_folder)
+file(TO_NATIVE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/generated" generated_files)
+file(TO_NATIVE_PATH "${generated_files}/attributes.winmd" attributes_winmd)
+file(TO_NATIVE_PATH "${generated_files}/component.winmd" component_winmd)
+file(TO_NATIVE_PATH "${project_folder}/compare/out/winrt/Component.h" compare_component_h)
```

Comment:
What is strategy behind checking in all the generated compare* files?  I'm missing what they're being compare with.

Reply:
The compare folder is what cpp.exe generates for the component so that any changes to the generated output can be compared with previous commits. That way we can easily tell when we've made a change to the code gen that affected the output and that may not necessarily trigger a unit test failure. This is exceedingly useful as I make changes to the compiler that I *think* should not impact code gen.

Reply:
Is it a temporary strategy?  How will it scale with?

  *   Automatic updating of the version string (e.g, base.h)
  *   Additional test collateral

Maybe a better long-term strategy, as we release binaries, is to run the last version of the tool and compare outputs without checking them into source control.

Reply:
* Version string should not be checked in so that shouldn't affect the output. 
* Additional tests would intentionally produce new code to compare. 

Once code gen is stable - I've reached parity with C+/WinrT - and tests provide reasonable code coverage, changes to the compare folder should be minimal/none, but provide a very convenient early warning system in the dev inner loop that something has changed. Pulling it out of the repo makes it an outer loop step that must be manually triggered and compared with an external tool. We tried that for the last rev of C++/WinRT and it was very inconvenient.

Reply:
version string will soon be injected into the tool by the nightly build, to uniquely stamp the exe.  that will cause diffs every night, yes?

---
In reply to: [229060580](https://github.com/Microsoft/xlang/pull/60#discussion_r229060580) [](ancestors = 229060580)

Reply:
Yes, but the build need not commit that version string so it should not affect the commit history.

Reply:
I don't see how that would make a difference.  The build will produce a versioned cpp.exe, which in turn will produced version headers, even with the version number being volatile (not committed).

Reply:
That's of no consequence since the build won't be committing its changed.

Reply:
offline conversation cleared things up - this is a visual check only (not an automated diff)